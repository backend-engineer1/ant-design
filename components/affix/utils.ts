import addEventListener from 'rc-util/lib/Dom/addEventListener';
import getScroll from '../_util/getScroll';
import Affix from './';

// ======================== Observer ========================
const TRIGGER_EVENTS = [
  'resize',
  'scroll',
  'touchstart',
  'touchmove',
  'touchend',
  'pageshow',
  'load',
];

interface ObserverEntity {
  target: HTMLElement | Window;
  affixList: Affix[];
  eventHandlers: { [eventName: string]: any };
}

let observerEntities: ObserverEntity[] = [];

export function addObserveTarget(target: HTMLElement | Window | null, affix: Affix): void {
  console.log('???', target);
  if (!target) return;

  let entity: ObserverEntity | undefined = observerEntities.find(item => item.target === target);

  if (entity) {
    entity.affixList.push(affix);
  } else {
    entity = {
      target,
      affixList: [affix],
      eventHandlers: {},
    };
    observerEntities.push(entity);

    // Add listener
    TRIGGER_EVENTS.forEach(eventName => {
      entity!.eventHandlers[eventName] = addEventListener(target, eventName, (event: Event) => {
        entity!.affixList.forEach(affix => {
          affix.updatePosition(event);
        });
      });
    });
  }
}

export function removeObserveTarget(affix: Affix): void {
  const observerEntity = observerEntities.find(oriObserverEntity => {
    const hasAffix = oriObserverEntity.affixList.some(item => item === affix);
    if (hasAffix) {
      oriObserverEntity.affixList = oriObserverEntity.affixList.filter(item => item !== affix);
    }
    return hasAffix;
  });

  if (observerEntity && observerEntity.affixList.length === 0) {
    observerEntities = observerEntities.filter(item => item !== observerEntity);

    // Remove listener
    TRIGGER_EVENTS.forEach(eventName => {
      const handler = observerEntity.eventHandlers[eventName];
      if (handler && handler.remove) {
        handler.remove();
      }
    });
  }
}

export function getTargetRect(target: HTMLElement | Window | null): ClientRect {
  return target !== window
    ? (target as HTMLElement).getBoundingClientRect()
    : ({ top: 0, bottom: window.innerHeight } as ClientRect);
}

export function getOffset(element: HTMLElement, target: HTMLElement | Window | null) {
  const elemRect = element.getBoundingClientRect();
  const targetRect = getTargetRect(target);

  const scrollTop = getScroll(target, true);
  const scrollLeft = getScroll(target, false);

  const docElem = window.document.body;
  const clientTop = docElem.clientTop || 0;
  const clientLeft = docElem.clientLeft || 0;

  return {
    top: elemRect.top - targetRect.top + scrollTop - clientTop,
    left: elemRect.left - targetRect.left + scrollLeft - clientLeft,
    width: elemRect.width,
    height: elemRect.height,
  };
}
