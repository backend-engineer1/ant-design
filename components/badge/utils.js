import React from 'react';

export function toArrayChildren(children) {
  const ret = [];
  React.Children.forEach(children, (c) => {
    ret.push(c);
  });
  return ret;
}

export function getPartNumber(num) {
  if (!num) {
    return null;
  }
  const countStr = num.toString();
  const obj = {};
  for (let i = 0; i < countStr.length; i++) {
    obj[countStr.length - 1 - i] = Number(countStr[i]);
  }
  return obj;
}

export function getTranslateY(differ, data, _data, height, i) {
  let translateY = 0;
  if (!differ) {
    //不想插入40个，改变要滚到的距离；
    if (_data[i + '_add']) {
      return -(data[i] + 20) * height;
    }
    if (_data[i + '_rem']) {
      return -data[i] * height;
    }
    return false;
  }
  //判断相差的位数来驱动数字；差个位滚动一格，差10的部分及以上个位滚动一周；
  const countToString = Math.abs(differ).toString();
  const countLength = countToString.length;
  const on = _data[i];
  const to = data[i];
  if (differ > 0) {
    if (countLength - 1 > i) {
      translateY = -(to - (to - on)) * height;
      //on大于to且differ大于10，如9->0,需要设计滚动到的位置＋10;
      if (on > to) {
        data[i + '_add'] = true;
      }
    } else if (countLength - 1 < i && to === on) {
      //参数不变动；
      translateY = null;
    } else if (typeof on === 'undefined') {
      //新增加入时设为0；
      translateY = -10 * height;
    } else {
      //如果开始大于到达，到达（to）加10；
      const _to = on > to ? to + 10 : to;
      translateY = -(to - (_to - on) + 10) * height;
    }
  } else {
    if (countLength - 1 > i) {
      translateY = -(to + (on - to) + 20) * height;
      //同上，differ大于10时，且to在于on
      if (to > on) {
        data[i + '_rem'] = true;
      }
    } else if (countLength - 1 < i && to === on) {
      translateY = null;
    } else {
      const _on = on < to ? on + 10 : on;
      translateY = -(to + (_on - to) + 10) * height;
    }
  }
  return translateY;
}
