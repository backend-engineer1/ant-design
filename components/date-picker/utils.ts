import * as moment from 'moment';

export function formatDate(
  value: moment.Moment | undefined | null,
  format: string | string[],
): string {
  if (!value) {
    return '';
  }
  if (Array.isArray(format)) {
    format = format[0];
  }
  return value.format(format);
}
