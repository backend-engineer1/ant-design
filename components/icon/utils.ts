// These props make sure that the SVG behaviours like general text.
// Reference: https://blog.prototypr.io/align-svg-icons-to-text-and-say-goodbye-to-font-icons-d44b3d7b26b4
export const svgBaseProps = {
  width: '1em',
  height: '1em',
  fill: 'currentColor',
  ['aria-hidden']: 'true',
};
