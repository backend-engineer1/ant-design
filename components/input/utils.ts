import type { InputProps } from './Input';

// eslint-disable-next-line import/prefer-default-export
export function hasPrefixSuffix(props: InputProps) {
  return !!(props.prefix || props.suffix || props.allowClear);
}
