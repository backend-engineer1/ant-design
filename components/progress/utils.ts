import { presetPrimaryColors } from '@ant-design/colors';
import type { CircleProps } from './Circle';
import type { ProgressProps } from './progress';
import warning from '../_util/warning';

export function validProgress(progress?: number) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}

export function getSuccessPercent({ success, successPercent }: ProgressProps) {
  let percent = successPercent;
  /** @deprecated Use `percent` instead */
  if (success && 'progress' in success) {
    warning(
      false,
      'Progress',
      '`success.progress` is deprecated. Please use `success.percent` instead.',
    );
    percent = success.progress;
  }
  if (success && 'percent' in success) {
    percent = success.percent;
  }
  return percent;
}

export const getPercentage = ({ percent, success, successPercent }: ProgressProps) => {
  const realSuccessPercent = validProgress(getSuccessPercent({ success, successPercent }));
  return [realSuccessPercent, validProgress(validProgress(percent) - realSuccessPercent)];
};

export const getStrokeColor = ({ success = {}, strokeColor }: Partial<CircleProps>): (
  | string
  | Record<PropertyKey, string>
)[] => {
  const { strokeColor: successColor } = success;
  return [successColor || presetPrimaryColors.green, strokeColor || null!];
};

export const getSize = (
  size: ProgressProps['size'],
  type: ProgressProps['type'] | 'step',
  extra?: {
    steps?: number;
    strokeWidth?: number;
  },
): [number, number] => {
  let width: number = -1;
  let height: number = -1;
  if (type === 'step') {
    const steps = extra!.steps!;
    const strokeWidth = extra!.strokeWidth!;
    if (typeof size === 'string' || typeof size === 'undefined') {
      width = size === 'small' ? 2 : 14;
      height = strokeWidth ?? 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = 14, height = 8] = size;
    }
    width *= steps;
  } else if (type === 'line') {
    const strokeWidth = extra?.strokeWidth;
    if (typeof size === 'string' || typeof size === 'undefined') {
      height = strokeWidth || (size === 'small' ? 6 : 8);
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = -1, height = 8] = size;
    }
  } else if (type === 'circle' || type === 'dashboard') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      [width, height] = size === 'small' ? [60, 60] : [120, 120];
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      if (process.env.NODE_ENV !== 'production') {
        warning(
          false,
          'Progress',
          'Type "circle" and "dashbord" do not accept array as `size`, please use number or preset size instead.',
        );
      }

      width = size[0] ?? size[1] ?? 120;
      height = size[0] ?? size[1] ?? 120;
    }
  }
  return [width, height];
};
