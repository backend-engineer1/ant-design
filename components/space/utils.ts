import type { SpaceSize } from '.';
import type { SizeType } from '../config-provider/SizeContext';

export function isPresetSize(size: SpaceSize): size is SizeType {
  return ['small', 'middle', 'large'].includes(size as string);
}

function isValidGapNumber(size: SpaceSize): size is number {
  if (!size) {
    // The case of size = 0 is deliberately excluded here, because the default value of the gap attribute in CSS is 0, so if the user passes 0 in, we can directly ignore it.
    return false;
  }
  return typeof size === 'number' && !Number.isNaN(size);
}

export const getRealSize = (
  sizeMap: Record<NonNullable<SpaceSize>, number>,
  value: SpaceSize,
): number => {
  if (isPresetSize(value)) {
    return sizeMap[value!];
  }
  if (isValidGapNumber(value)) {
    return value;
  }
  return 0;
};
