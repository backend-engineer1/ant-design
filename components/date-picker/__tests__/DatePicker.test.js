import React from 'react';
import { mount } from 'enzyme';
import moment from 'moment';
import MockDate from 'mockdate';
import DatePicker from '..';
import { selectDate } from './utils';
import focusTest from '../../../tests/shared/focusTest';

describe('DatePicker', () => {
  focusTest(DatePicker);

  beforeEach(() => {
    MockDate.set(new Date('2016-11-22').getTime() + (new Date().getTimezoneOffset() * 60 * 1000));
  });

  afterEach(() => {
    MockDate.reset();
  });

  it('prop locale should works', () => {
    const locale = {
      lang: {
        placeholder: 'Избери дата',
        rangePlaceholder: [
          'Начална дата',
          'Крайна дата',
        ],
        today: 'Днес',
        now: 'Сега',
        backToToday: 'Към днес',
        ok: 'Добре',
        clear: 'Изчистване',
        month: 'Месец',
        year: 'Година',
        timeSelect: 'Избор на час',
        dateSelect: 'Избор на дата',
        monthSelect: 'Избор на месец',
        yearSelect: 'Избор на година',
        decadeSelect: 'Десетилетие',
        previousMonth: 'Предишен месец (PageUp)',
        nextMonth: 'Следващ месец (PageDown)',
        previousYear: 'Последна година (Control + left)',
        nextYear: 'Следваща година (Control + right)',
        previousDecade: 'Предишно десетилетие',
        nextDecade: 'Следващо десетилетие',
        previousCentury: 'Последен век',
        nextCentury: 'Следващ век',
        yearFormat: 'YYYY',
        dateFormat: 'D M YYYY',
        dayFormat: 'D',
        dateTimeFormat: 'D M YYYY HH:mm:ss',
        monthBeforeYear: true,
      },
      timePickerLocale: {
        placeholder: 'Избор на час',
      },
    };
    const birthday = moment('2000-01-01', 'YYYY-MM-DD');
    const wrapper = mount(
      <DatePicker open locale={locale} value={birthday} />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  // Fix https://github.com/ant-design/ant-design/issues/8885
  it('control value after panel closed', () => {
    class Test extends React.Component {
      state = {
        cleared: false,
        value: moment(),
      }

      onChange = (value) => {
        let cleared = this.state.cleared;

        if (cleared) {
          value = moment(moment(value).format('YYYY-MM-DD 12:12:12'));
          cleared = false;
        }

        if (!value) {
          cleared = true;
        }

        this.setState({ value, cleared });
      }

      render() {
        return (
          <div>
            <DatePicker
              showTime
              value={this.state.value}
              format="YYYY-MM-DD HH:mm:ss"
              onChange={this.onChange}
            />
          </div>
        );
      }
    }

    const wrapper = mount(<Test />);
    // clear input
    wrapper.find('.ant-calendar-picker-clear').hostNodes().simulate('click');
    // open panel
    wrapper.find('.ant-calendar-picker-input').simulate('click');
    // select a date
    selectDate(wrapper, moment('2016-11-13'));
    expect(wrapper.find('.ant-calendar-input').getDOMNode().value).toBe('2016-11-13 12:12:12');
    selectDate(wrapper, moment('2016-11-14'));
    expect(wrapper.find('.ant-calendar-input').getDOMNode().value).toBe('2016-11-14 12:12:12');
  });
});
