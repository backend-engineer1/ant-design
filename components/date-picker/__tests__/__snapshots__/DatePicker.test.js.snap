// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker prop locale should works 1`] = `
<span
  class="ant-calendar-picker"
>
  <div
    style=""
  >
    <input
      class="ant-calendar-picker-input ant-input"
      placeholder="Избери дата"
      readonly=""
      value="2000-01-01"
    />
    <i
      class="anticon ant-calendar-picker-clear"
    />
    <i
      class="anticon ant-calendar-picker-icon"
    >
      <svg
        aria-hidden="true"
        data-icon="calendar"
        fill="currentColor"
        height="1em"
        viewBox="0 0 1024 1024"
        width="1em"
      >
        <path
          class="st0"
          d="M184 460h656v380H184zM712 256v48c0 4.4-3.6 8-8 8h-56c-4.4 0-8-3.6-8-8v-48H384v48c0 4.4-3.6 8-8 8h-56c-4.4 0-8-3.6-8-8v-48H184v136h656V256H712z"
        />
        <path
          d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
        />
      </svg>
    </i>
  </div>
  <div>
    <div
      class="ant-calendar-picker-container ant-calendar-picker-container-placement-bottomLeft slide-up-appear"
      style="left: -999px; top: -1002px;"
    >
      <div
        class="ant-calendar"
        tabindex="0"
      >
        <div
          class="ant-calendar-panel"
        >
          <div
            class="ant-calendar-input-wrap"
          >
            <div
              class="ant-calendar-date-input-wrap"
            >
              <input
                class="ant-calendar-input "
                placeholder="Избери дата"
                value="2000-01-01"
              />
            </div>
            <a
              class="ant-calendar-clear-btn"
              role="button"
              title="Изчистване"
            />
          </div>
          <div
            class="ant-calendar-date-panel"
          >
            <div
              class="ant-calendar-header"
            >
              <div
                style="position: relative;"
              >
                <a
                  class="ant-calendar-prev-year-btn"
                  role="button"
                  title="Последна година (Control + left)"
                />
                <a
                  class="ant-calendar-prev-month-btn"
                  role="button"
                  title="Предишен месец (PageUp)"
                />
                <span
                  class="ant-calendar-my-select"
                >
                  <a
                    class="ant-calendar-month-select"
                    role="button"
                    title="Избор на месец"
                  >
                    Jan
                  </a>
                  <a
                    class="ant-calendar-year-select"
                    role="button"
                    title="Избор на година"
                  >
                    2000
                  </a>
                </span>
                <a
                  class="ant-calendar-next-month-btn"
                  title="Следващ месец (PageDown)"
                />
                <a
                  class="ant-calendar-next-year-btn"
                  title="Следваща година (Control + right)"
                />
              </div>
            </div>
            <div
              class="ant-calendar-body"
            >
              <table
                cellspacing="0"
                class="ant-calendar-table"
                role="grid"
              >
                <thead>
                  <tr
                    role="row"
                  >
                    <th
                      class="ant-calendar-column-header"
                      role="columnheader"
                      title="Sun"
                    >
                      <span
                        class="ant-calendar-column-header-inner"
                      >
                        Su
                      </span>
                    </th>
                    <th
                      class="ant-calendar-column-header"
                      role="columnheader"
                      title="Mon"
                    >
                      <span
                        class="ant-calendar-column-header-inner"
                      >
                        Mo
                      </span>
                    </th>
                    <th
                      class="ant-calendar-column-header"
                      role="columnheader"
                      title="Tue"
                    >
                      <span
                        class="ant-calendar-column-header-inner"
                      >
                        Tu
                      </span>
                    </th>
                    <th
                      class="ant-calendar-column-header"
                      role="columnheader"
                      title="Wed"
                    >
                      <span
                        class="ant-calendar-column-header-inner"
                      >
                        We
                      </span>
                    </th>
                    <th
                      class="ant-calendar-column-header"
                      role="columnheader"
                      title="Thu"
                    >
                      <span
                        class="ant-calendar-column-header-inner"
                      >
                        Th
                      </span>
                    </th>
                    <th
                      class="ant-calendar-column-header"
                      role="columnheader"
                      title="Fri"
                    >
                      <span
                        class="ant-calendar-column-header-inner"
                      >
                        Fr
                      </span>
                    </th>
                    <th
                      class="ant-calendar-column-header"
                      role="columnheader"
                      title="Sat"
                    >
                      <span
                        class="ant-calendar-column-header-inner"
                      >
                        Sa
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="ant-calendar-tbody"
                >
                  <tr
                    class="ant-calendar-active-week"
                    role="row"
                  >
                    <td
                      class="ant-calendar-cell ant-calendar-last-month-cell"
                      role="gridcell"
                      title="December 26, 1999"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        26
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-last-month-cell"
                      role="gridcell"
                      title="December 27, 1999"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        27
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-last-month-cell"
                      role="gridcell"
                      title="December 28, 1999"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        28
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-last-month-cell"
                      role="gridcell"
                      title="December 29, 1999"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        29
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-last-month-cell"
                      role="gridcell"
                      title="December 30, 1999"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        30
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-last-month-cell ant-calendar-last-day-of-month"
                      role="gridcell"
                      title="December 31, 1999"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        31
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-selected-date ant-calendar-selected-day"
                      role="gridcell"
                      title="January 1, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="true"
                        class="ant-calendar-date"
                      >
                        1
                      </div>
                    </td>
                  </tr>
                  <tr
                    class=""
                    role="row"
                  >
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 2, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        2
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 3, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        3
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 4, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        4
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 5, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        5
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 6, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        6
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 7, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        7
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 8, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        8
                      </div>
                    </td>
                  </tr>
                  <tr
                    class=""
                    role="row"
                  >
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 9, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        9
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 10, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        10
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 11, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        11
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 12, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        12
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 13, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        13
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 14, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        14
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 15, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        15
                      </div>
                    </td>
                  </tr>
                  <tr
                    class=""
                    role="row"
                  >
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 16, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        16
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 17, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        17
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 18, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        18
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 19, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        19
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 20, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        20
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 21, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        21
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 22, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        22
                      </div>
                    </td>
                  </tr>
                  <tr
                    class=""
                    role="row"
                  >
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 23, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        23
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 24, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        24
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 25, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        25
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 26, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        26
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 27, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        27
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 28, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        28
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 29, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        29
                      </div>
                    </td>
                  </tr>
                  <tr
                    class=""
                    role="row"
                  >
                    <td
                      class="ant-calendar-cell"
                      role="gridcell"
                      title="January 30, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        30
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-last-day-of-month"
                      role="gridcell"
                      title="January 31, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        31
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-next-month-btn-day"
                      role="gridcell"
                      title="February 1, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        1
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-next-month-btn-day"
                      role="gridcell"
                      title="February 2, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        2
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-next-month-btn-day"
                      role="gridcell"
                      title="February 3, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        3
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-next-month-btn-day"
                      role="gridcell"
                      title="February 4, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        4
                      </div>
                    </td>
                    <td
                      class="ant-calendar-cell ant-calendar-next-month-btn-day"
                      role="gridcell"
                      title="February 5, 2000"
                    >
                      <div
                        aria-disabled="false"
                        aria-selected="false"
                        class="ant-calendar-date"
                      >
                        5
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div
              class="ant-calendar-footer"
            >
              <span
                class="ant-calendar-footer-btn"
              >
                <a
                  class="ant-calendar-today-btn "
                  role="button"
                  title="November 22, 2016"
                >
                  Днес
                </a>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</span>
`;
